handler/core/src/main/java/io/knotx/fragments/handler/action/CircuitBreakerActionFactory.java

```diff
@@ -46,46 +59,94 @@ public String getName() {
 
-    CircuitBreakerAction(CircuitBreaker circuitBreaker, Action doAction) {
+    CircuitBreakerAction(CircuitBreaker circuitBreaker, Action doAction, String alias,
+        ActionLogLevel actionLogLevel, String errorTransition) {
       this.circuitBreaker = circuitBreaker;
       this.doAction = doAction;
+      this.alias = alias;
+      this.actionLogLevel = actionLogLevel;
+      this.errorTransition = errorTransition;
     }
 
     @Override
     public void apply(FragmentContext fragmentContext,
         Handler<AsyncResult<FragmentResult>> resultHandler) {
+      AtomicInteger counter = new AtomicInteger();
+      ActionLogger actionLogger = ActionLogger.create(alias, actionLogLevel);
+
       circuitBreaker.executeWithFallback(
-          f -> doAction.apply(fragmentContext,
-              result -> {
-                if (result.succeeded()) {
-                  f.complete(result.result());
-                } else {
-                  f.fail(result.cause());
-                }
-              }),
-          v -> {
-            Fragment fragment = fragmentContext.getFragment();
-            return new FragmentResult(fragment, FALLBACK_TRANSITION);
-          }
+          promise -> executeCommand(promise, fragmentContext, counter, actionLogger),
+          throwable -> handleFallback(fragmentContext, throwable, counter, actionLogger)
       ).setHandler(resultHandler);
     }
+
+    private void executeCommand(Promise<FragmentResult> f, FragmentContext fragmentContext,
+        AtomicInteger counter, ActionLogger actionLogger) {
+      int index = counter.addAndGet(1);
+      long startTime = now().toEpochMilli();
+      doAction.apply(fragmentContext,
+          result -> doActionResultHandler(f, result.result(), index, startTime, actionLogger));
+    }
+
+    private void doActionResultHandler(Promise<FragmentResult> promise,
+        FragmentResult result, int invocation, long startTime, ActionLogger actionLogger) {
+      if (isError(result)) {
+        handleFail(promise, result, invocation, startTime, actionLogger);
+      } else {
+        handleSuccess(promise, result, invocation, startTime, actionLogger);
+      }
+    }
+
+    private boolean isError(FragmentResult result){
+      return errorTransition.equals(result.getTransition());
+    }
+    private void handleFail(Promise<FragmentResult> f, FragmentResult result,
+        int invocation, long startTime, ActionLogger actionLogger) {
+      actionLogger.failedDoActionLog(executionTime(startTime),  result.getNodeLog());
+      f.fail(new DoActionExecuteException(format("Action end up %s transition", errorTransition)));
+    }
+
+    private static long executionTime(long startTime){
+      return now().toEpochMilli() - startTime;
+    }
+
+    private void handleSuccess(Promise<FragmentResult> f, FragmentResult result,
+        int invocation, long startTime, ActionLogger actionLogger) {
+      actionLogger.info("invocationCount", valueOf(invocation));
```

Comment:
What if we have:
- first invocation, cb timeout
- second invocation, update `invocationCount` (2)
- first invocation finished, update `invocationCount` (1)

Is it possible with Vert.x Event Loop? @Skejven what is your opinion?

Reply:
@marcinkp Do we have unit test for that?

Reply:
Not sure how it works with CB. Will it still hold some reference after timeout? Good idea to check it out with the test.